lib/commons/dom/is-visible.js

```diff
@@ -13,14 +15,187 @@ const clipRegex = /rect\s*\(([0-9]+)px,?\s*([0-9]+)px,?\s*([0-9]+)px,?\s*([0-9]+
 const clipPathRegex = /(\w+)\((\d+)/;
 
 /**
- * Determines if an element is hidden with a clip or clip-path technique
- * @method isClipped
+ * Determine whether an element is visible
+ * @method isVisible
  * @memberof axe.commons.dom
- * @private
- * @param  {CSSStyleDeclaration} style Computed style
- * @return {Boolean}
+ * @instance
+ * @param {HTMLElement} el The HTMLElement
+ * @param {Boolean} screenReader When provided, will evaluate visibility from the perspective of a screen reader
+ * @param {Boolean} recursed
+ * @return {Boolean} The element's visibilty status
+ */
+export default function isVisible(el, screenReader, recursed) {
+  assert(el, 'Cannot determine if element is visible for non-DOM nodes');
+  const vNode = el instanceof AbstractVirtualNode ? el : getNodeFromTree(el);
+  el = vNode ? vNode.actualNode : el;
+
+  const cacheName = '_isVisible' + (screenReader ? 'ScreenReader' : '');
+  const { DOCUMENT_NODE } = window.Node ?? {};
+
+  // if we don't have a virtual node that means we've
+  // recursed up the parent tree to a node that is
+  // outside of the context. we still need to know
+  // if the node and parents are hidden so need to
+  // work with both virtual and non-virtual nodes
+  const nodeType = vNode ? vNode.props.nodeType : el.nodeType;
+  const nodeName = vNode ? vNode.props.nodeName : el.nodeName.toLowerCase();
+
+  if (vNode && typeof vNode[cacheName] !== 'undefined') {
+    return vNode[cacheName];
+  }
+
+  if (nodeType === DOCUMENT_NODE) {
+    return true;
+  }
+
+  const hiddenMethods = [elementHidden];
+  let style;
+  if (screenReader) {
+    hiddenMethods.push(ariaHidden);
+  }
+
+  if (el) {
+    style = window.getComputedStyle(el);
+
+    /**
+     * check visibility of `AREA`
+     * Note:
+     * Firefox's user-agent always sets `AREA` element to `display:none`
+     * hence excluding the edge case, for visibility computation
+     */
+    if (nodeName === 'area') {
+      return isAreaVisible(el, screenReader, recursed);
+    }
+
+    hiddenMethods.push(displayHidden);
+
+    if (!screenReader) {
+      hiddenMethods.push(
+        opacityHidden,
+        clipHidden,
+        scrollHidden,
+        overflowHidden
+      );
+    }
+    if (!recursed) {
+      hiddenMethods.push(visibilityHidden);
+    }
+    if (!screenReader && !recursed) {
+      hiddenMethods.push(offscreenHidden);
+    }
+  }
+
+  if (
+    hiddenMethods.some(method =>
+      method({
+        el,
+        vNode,
+        nodeName,
+        style
+      })
+    )
+  ) {
+    return false;
+  }
+
+  if (nodeName === 'html') {
```

Comment:
`getComposedParent` doesn't return the `#document` node, so we need to stop recursing at the `html` element. However we have a test for making sure `#document` is visible, so had to still include the `nodeType` check for `#document`.